src/hotspot/share/interpreter/bytecodeTracer.cpp

```diff
@@ -396,8 +388,78 @@ void BytecodePrinter::print_field_or_method(int orig_i, int i, outputStream* st)
     }
     st->print_cr(" %d <%s%s%s>", i, name->as_C_string(), sep, signature->as_C_string());
   }
+
+  if (ClassPrinter::has_mode(_flags, ClassPrinter::PRINT_DYNAMIC) &&
+      (tag.is_dynamic_constant() || tag.is_invoke_dynamic())) {
+    print_dynamic(orig_i, i, tag, st);
+  }
 }
 
+void BytecodePrinter::print_dynamic(int orig_i, int i, constantTag tag, outputStream* st) {
+  ConstantPool* constants = method()->constants();
+  int bsm = constants->bootstrap_method_ref_index_at(i);
+  const char* ref_kind = "";
+  switch (constants->method_handle_ref_kind_at(bsm)) {
+  case JVM_REF_getField         : ref_kind = "REF_getField"; break;
+  case JVM_REF_getStatic        : ref_kind = "REF_getStatic"; break;
+  case JVM_REF_putField         : ref_kind = "REF_putField"; break;
+  case JVM_REF_putStatic        : ref_kind = "REF_putStatic"; break;
+  case JVM_REF_invokeVirtual    : ref_kind = "REF_invokeVirtual"; break;
+  case JVM_REF_invokeStatic     : ref_kind = "REF_invokeStatic"; break;
+  case JVM_REF_invokeSpecial    : ref_kind = "REF_invokeSpecial"; break;
+  case JVM_REF_newInvokeSpecial : ref_kind = "REF_newInvokeSpecial"; break;
+  case JVM_REF_invokeInterface  : ref_kind = "REF_invokeInterface"; break;
+  default                       : ShouldNotReachHere();
+  }
+  st->print("  BSM: %s", ref_kind);
+  print_field_or_method(-i, constants->method_handle_index_at(bsm), st);
+  int argc = constants->bootstrap_argument_count_at(i);
+  st->print("  arguments[%d] = {", argc);
+  if (argc > 0) {
+    st->cr();
+    for (int arg_i = 0; arg_i < argc; arg_i++) {
+      int arg = constants->bootstrap_argument_index_at(i, arg_i);
+      st->print("    ");
+      print_constant_nocheck(arg, st);
+    }
+  }
+  st->print_cr("  }");
+  if (tag.is_invoke_dynamic()) {
+    int indy_index = orig_i;
+    int cpc_index = constants->invokedynamic_cp_cache_index(indy_index);
+    print_cpcache_entry(cpc_index, st);
+  } else {
+    // TODO: print info for tag.is_dynamic_constant()
+  }
+}
```

Comment:
This should be a function in ConstantPool::print()

Reply:
This block of code works on an `indy_index`, which is stored only inside the bytecode stream (it's the rewritten index of the [invokedynamic bytecode](https://docs.oracle.com/javase/specs/jvms/se17/html/jvms-6.html#jvms-6.5.invokedynamic). `indy_index` is not store anywhere inside the `ConstantPool`, so I can't find a good place to print this info in the printing functions of `ConstantPool`.

Note that you can have multiple `invokedynamic` bytecodes that use the same `JVM_CONSTANT_InvokeDynamic` entry in a `ConstantPool`, resulting in a different  `indy_index` for each call site. Therefore, this per-callsite information cannot be printed as part of the `JVM_CONSTANT_InvokeDynamic` entry

Reply:
ok